man/powder.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "POWDER" "1" "August 2012" "" ""
.
.SH "NAME"
\fBpowder\fR \- manage Pow Rack server
.
.SH "SYNOPSIS"
\fBpowder\fR
.
.SH "DESCRIPTION"
Command\-line management utility for 37signals\' Pow zero\-configuration Rack server for OS X (http://pow\.cx/)\.
.
.SH "HOMEPAGE"
https://github\.com/Rodreegez/powder
.
.SH "EXAMPLES"
.
.SS "Display usage information"
\fB$ powder [\-h|help]\fR
.
.br
=> Display usage information
.
.br
# Lists name and brief descriptions of the tasks available
.
.SS "Linking applications in Pow"
powder will attempt to read \.powder, which names a default symlink for the current project
.
.P
\fB$ powder link\fR
.
.br
=> Link the current dir to ~/\.pow/\fIcurrent_directory\fR
.
.P
\fB$ powder link [bacon]\fR
.
.br
=> Link the current dir to ~/\.pow/bacon
.
.br
=> Create \.powder, contents bacon
.
.P
\fB$ powder link [bacon] \-\-no\-create\fR
.
.br
=> Link the current dir to ~/\.pow/bacon
.
.P
\fB$ powder link [bacon] \-\-force\fR
.
.br
=> Remove the current pow symlink, and \.powder
.
.br
=> Link the current dir to ~/\.pow/bacon
.
.br
=> Create \.powder, contents bacon
.
.P
\fB$ powder default\fR => Link the current dir to ~/\.pow/default # Serve this directory for all unhandled domains
.
.P
For both forms of link, if the current directory doesn\'t look like an app that can be powed it will offer to download a basic \fBconfig\.ru\fR for Rails 2
.
.SS "Unlinking applications"
\fB$ powder unlink\fR
.
.br
=> Unlink current_dir or the symlink defined in \.powder
.
.P
\fB$ powder unlink bacon\fR
.
.br
=> Unlink bacon
.
.P
\fB$ powder cleanup\fR
.
.br
=> remove all invalid symbolic links
.
.SS "Working with Pow"
\fB$ powder config\fR
.
.br
=> Get Pow\'s current configuration information
.
.P
\fB$ powder list\fR
.
.br
=> List all the current apps linked in ~/\.pow
.
.br
# aliased as \fBpowder \-l\fR
.
.P
\fB$ powder status\fR
.
.br
=> Get Pow\'s current status information
.
.P
\fB$ powder version\fR
.
.br
=> Returns the current powder version
.
.br
# aliased as \fBpowder \-v\fR
.
.SS "Working with applications"
\fB$ powder open\fR
.
.br
=> Opens the pow link in a browser
.
.br
# aliased as powder \-o
.
.P
\fB$ powder open [bacon]\fR
.
.br
=> Opens http://bacon\.test in a browser
.
.br
# if you have set up alternative top level domains in \.powconfig,
.
.br
# then the first listed domain will be opened\.
.
.P
\fB$ powder open \-\-browser Safari\fR => Opens the pow link in a specific browser (in this case, Safari) # Also aliased as \-b
.
.P
\fB$ powder open \-b \'Google Chrome\'\fR => Opens the pow link with browsers with more than one word
.
.P
\fB$ powder open \-\-path home\fR => Opens the pow link with a path appended (http://bacon.test/home) # Also aliased as \-p
.
.P
# Should also works with all the other \'open\' options: \fB$ powder open bacon \-b Safari\fR \fB$ powder open \-\-xip \-b Firefox\fR \fB$ powder \-o \-x \-b \'Google Chrome\'\fR
.
.SS "Managing application restarts"
\fB$ powder restart\fR
.
.br
=> Restart the current app
.
.br
# aliased as \fBpowder \-r\fR
.
.P
\fB$ powder always_restart\fR
.
.br
=> Always restart the current app
.
.br
# aliased as powder \-a
.
.P
\fB$ powder no_restarts\fR
.
.br
=> don\'t do any automatic restarting of the current app
.
.SS "Working with logs"
\fB$ powder applog\fR
.
.br
=> tail the log of the current app
.
.P
\fB$ powder log\fR
.
.br
=> Tails the pow log\.
.
.br
# Not the application log, but the pow log, available at
.
.br
# ~/Library/Logs/Pow/apps/#{app\-directory}\.log
.
.SS "Working with Environment Variables"
\fB$ powder env\fR => Displays your current custom pow environment variables # Pow reads environment varialbles from \.powenv
.
.P
\fB$ powder env_reset\fR => Deletes your \.powenv, removing all custom environment variables\.
.
.P
\fB$ powder env BACON chunky\fR => Pass an arbitrary environment variable to pow, eg, ENV["BACON"] = "chunky" # Remove an ENV by passing in no value, eg: powder env BACON # If you already have a \.gitignore, the newly created \.powenv will also be ignored automatically\.
.
.P
\fB$ powder [production|development|test]\fR => Run your Rails app as Production # aliased as powder [prod|dev] # This is a wrapper for powder env RAILS_ENV \.\.\.
.
.SS "Installing and uninstalling Pow"
\fB$ powder install\fR
.
.br
=> Installs pow server
.
.br
# (I know, "curl get\.pow\.cx | sh" isn\'t hard, but this is \fIeven\fR easier)
.
.P
\fB$ powder uninstall\fR
.
.br
=> Uninstalls pow server
.
.P
\fB$ powder update\fR
.
.br
=> Updates pow server
.
.br
# Really this is just an alias to powder install, but it feels more natural
.
.br
# this way\.
.
.SS "Enabling and Disabling Pow"
\fB$ powder up\fR
.
.br
=> Enable Pow
.
.P
\fB$ powder down\fR
.
.br
=> Disable Pow
.
.P
\fB$ powder debug\fR => Opens a debug shell with your application environment
.
.SH "AUTHOR"
Built by rodreegez \fIhttps://github\.com/Rodreegez\fR and philnash \fIhttps://github\.com/philnash\fR\.


CHANGELOG
=== 0.1.2 / TBA

* 1 minor change
  * add log command - https://github.com/Rodreegez/powder/pull/8

=== 0.1.1 / 2011-04-11

* 1 minor change
  * We won't try reading ~/.powconfig if it doesn't exist

=== 0.1.0 / 2011-04-11

* 1 Major change

  * The command has changed from pow to powder to allow pow to use the command
  (https://github.com/Rodreegez/powder/issues/5)
  
* 1 minor change

  * open and link now honour domains specified in .powconfig

=== 0.0.4 / 2011-04-08

* Early release with link, list, restart, remove, open, install and uninstall commands


lib/powder.rb
module Powder
  # Get the origin of the application link, whether it's a link to a
  # rack app, proxy to a port, or proxy to a URL.
  #
  # Mostly follows logic of pow configuration class last changed 2011-08-05,
  # which remains current as of pow v0.4.1 (and master on 2014-01-08)
  # http://github.com/37signals/pow/blob/36b91fe/lib/configuration.js#L161-L188
  def get_app_origin(app_link)
    case (File.stat(app_link).ftype)
    when "file"
      data = File.read(app_link)
      if data.length < 10 && data.to_i != 0
        "proxy port: #{data}"
      elsif url_match = (data.match %r{https?://\S+})
        "proxy URL: #{url_match.to_s}"
      else
        nil
      end
    when "directory"
      canonicalize_path(app_link)
    else
      nil
    end
  end

  # Prepare path to the level of accuracy we prefer for list display
  def canonicalize_path(path)
      path = File.readlink path if File.symlink? path
      path.sub(ENV['HOME'], '~')
  end
end
